mayachemtools/docs/modules/man3/PackageInfo.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PackageInfo 1"
.TH PackageInfo 1 "2018-05-15" "perl v5.22.4" "MayaChemTools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
PackageInfo
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
use PackageInfo;
.PP
use PackageInfo qw(:all);
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBPackageInfo\fR module provides the following functions:
.PP
GetPackageKeyValue, IsPackageKeyNameAvailable, SetPackageKeyValue
.PP
The functions to set and get package keyvalues not explicitly defined in this module
are implemented using Perl's \s-1AUTOLOAD\s0 functionality. These methods are generated
on-the-fly for a specified key:
.PP
.Vb 2
\&    Set<KeyName>(<KeyValue>);
\&    $KeyValue = Get<KeyName>();
.Ve
.PP
\&\fBPackageInfo\fR module provides functionality to retrieve information about MayaChemTools
package from PackagaInfo.csv which contains the following types key name and values:
.PP
.Vb 7
\&    "KeyName","KeyValue"
\&    "PackageName","MayaChemTools"
\&    "ReleaseDate","Oct 21, 2010"
\&    "VersionNumber","7.4"
\&    "DevSoftwareEnvironment","Cygwin on Windows XP"
\&    ... ...
\&    ... ...
.Ve
.SH "FUNCTIONS"
.IX Header "FUNCTIONS"
.IP "\fBGetPackageKeyValue\fR" 4
.IX Item "GetPackageKeyValue"
.Vb 1
\&    $KeyValue = GetPackageKeyValue($KeyName);
.Ve
.Sp
Returns \fBKeyValue\fR for a specified \fIKeyName\fR.
.IP "\fBIsPackageKeyNameAvailable\fR" 4
.IX Item "IsPackageKeyNameAvailable"
.Vb 1
\&    $Status = IsPackageKeyNameAvailable($KeyName);
.Ve
.Sp
Returns 1 or 0 based on whether \fIKeyName\fR is available in package info file.
.IP "\fBSetPackageKeyValue\fR" 4
.IX Item "SetPackageKeyValue"
.Vb 1
\&    SetPackageKeyValue($KeyName, $KeyValue);
.Ve
.Sp
Sets \fIKeyValue\fR for a \fIKeyName\fR. No data is written to package info file.
.SH "AUTHOR"
.IX Header "AUTHOR"
Manish Sud <msud@san.rr.com>
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (C) 2018 Manish Sud. All rights reserved.
.PP
This file is part of MayaChemTools.
.PP
MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the \s-1GNU\s0 Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.


scripts/mayachemtools/lib/PackageInfo.pm
package PackageInfo;
#
# File: PackageInfo.pm
# Author: Manish Sud <msud@san.rr.com>
#
# Copyright (C) 2018 Manish Sud. All rights reserved.
#
# This file is part of MayaChemTools.
#
# MayaChemTools is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option) any
# later version.
#
# MayaChemTools is distributed in the hope that it will be useful, but without
# any warranty; without even the implied warranty of merchantability of fitness
# for a particular purpose.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with MayaChemTools; if not, see <http://www.gnu.org/licenses/> or
# write to the Free Software Foundation Inc., 59 Temple Place, Suite 330,
# Boston, MA, 02111-1307, USA.
#

use strict;
use Exporter;
use Carp;
use Text::ParseWords;
use TextUtil;
use FileUtil;

use vars qw($AUTOLOAD @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);

@ISA = qw(Exporter);
@EXPORT = qw(GetPackageKeyValue SetPackageKeyValue IsPackageKeyNameAvailable);
@EXPORT_OK = qw();
%EXPORT_TAGS = (all  => [@EXPORT, @EXPORT_OK]);

#
# Load package data...
#
my(%PackageDataMap);
_LoadPackageData();

# Return value of a specific key...
#
sub GetPackageKeyValue {
  my($KeyName) = @_;

  return exists $PackageDataMap{$KeyName} ? $PackageDataMap{$KeyName} : 'Not Available';
}

# Set value of a specific key...
#
sub SetPackageKeyValue {
  my($KeyName, $KeyValue) = @_;

  $PackageDataMap{$KeyName} = $KeyValue;
}

# Check availability of a package key name...
#
sub IsPackageKeyNameAvailable {
  my($KeyName) = @_;

  return exists $PackageDataMap{$KeyName} ? 1 : 0;
}

# Implements Set<KeyName> and Get<KeyName> functions...
#
sub AUTOLOAD {
  my($KeyValue) = @_;
  my($PackageName, $FunctionName, $KeyName);

  ($PackageName, $FunctionName) = $AUTOLOAD =~ /^(.*?)::(.*?)$/;

  if (!($FunctionName =~ /^Get/ || $FunctionName =~ /^Set/)) {
    croak "Error: Can't locate function \"$FunctionName\" via package \"$PackageName\": This function is not automatically implemented by AUTOLOAD: Only Get<KeyName> and Set<KeyName> functions are implemented via AUTOLOAD...";
  }

  ($KeyName) = $FunctionName =~ /^[SG]et(.*?)$/;

  if ($FunctionName =~ /^Set/ && !defined($KeyValue)) {
    carp "Warning:  ${PackageName}::${FunctionName}: Didn't set value for key $KeyName: Key value for must be specified...\n";
    return undef;
  }

  if ($FunctionName =~ /^Get/) {
    return GetPackageKeyValue($KeyName);
  }
  elsif ($FunctionName =~ /^Set/) {
    return SetPackageKeyValue($KeyName, $KeyValue);
  }

}

# Load PackageInfo.csv files from <MayaChemTools>/lib directory...
#
sub _LoadPackageData {
  my($PackageDataFile, $MayaChemToolsLibDir, $Key, $Value, $Line, $InDelim, @LineWords);

  $MayaChemToolsLibDir = GetMayaChemToolsLibDirName();
  $PackageDataFile =  "$MayaChemToolsLibDir" . "/data/PackageInfo.csv";

  if (! -e "$PackageDataFile") {
    croak "Error: MayaChemTools package file, $PackageDataFile, is missing: Possible installation problems...";
  }

  #
  # Format:
  #
  # "Key","Value"
  # "PackageName","MayaChemTools"
  # ... ...
  #

  %PackageDataMap = ();
  $InDelim = "\,";

  open PACKAGEDATAFILE, "$PackageDataFile" or croak "Couldn't open $PackageDataFile: $! ...";

  # Skip lines up to column labels...
  LINE: while ($Line = GetTextLine(\*PACKAGEDATAFILE)) {
    if ($Line !~ /^#/) {
      last LINE;
    }
  }

  # Process key/value pairs...
  LINE: while ($Line = GetTextLine(\*PACKAGEDATAFILE)) {
    if ($Line =~ /^#/) {
      next LINE;
    }
    @LineWords = ();
    @LineWords = quotewords($InDelim, 0, $Line);
    if (@LineWords != 2) {
      croak "Error: The number of data fields, @LineWords, in $PackageDataFile must be 2.\nLine: $Line...";
    }
    ($Key, $Value) = @LineWords;

    if (exists $PackageDataMap{$Key}) {
      carp "Warning: Multiple entries for key, $Key, in $PackageDataFile. Ignoring current line.\nLine: $Line...";
      next LINE;
    }

    $PackageDataMap{$Key} = $Value;
  }
  close PACKAGEDATAFILE;
}

1;

__END__

=head1 NAME

PackageInfo

=head1 SYNOPSIS

use PackageInfo;

use PackageInfo qw(:all);

=head1 DESCRIPTION

B<PackageInfo> module provides the following functions:

GetPackageKeyValue, IsPackageKeyNameAvailable, SetPackageKeyValue

The functions to set and get package keyvalues not explicitly defined in this module
are implemented using Perl's AUTOLOAD functionality. These methods are generated
on-the-fly for a specified key:

    Set<KeyName>(<KeyValue>);
    $KeyValue = Get<KeyName>();

B<PackageInfo> module provides functionality to retrieve information about MayaChemTools
package from PackagaInfo.csv which contains the following types key name and values:

    "KeyName","KeyValue"
    "PackageName","MayaChemTools"
    "ReleaseDate","Oct 21, 2010"
    "VersionNumber","7.4"
    "DevSoftwareEnvironment","Cygwin on Windows XP"
    ... ...
    ... ...

=head1 FUNCTIONS

=over 4

=item B<GetPackageKeyValue>

    $KeyValue = GetPackageKeyValue($KeyName);

Returns B<KeyValue> for a specified I<KeyName>.

=item B<IsPackageKeyNameAvailable>

    $Status = IsPackageKeyNameAvailable($KeyName);

Returns 1 or 0 based on whether I<KeyName> is available in package info file.

=item B<SetPackageKeyValue>

    SetPackageKeyValue($KeyName, $KeyValue);

Sets I<KeyValue> for a I<KeyName>. No data is written to package info file.

=back

=head1 AUTHOR

Manish Sud <msud@san.rr.com>

=head1 COPYRIGHT

Copyright (C) 2018 Manish Sud. All rights reserved.

This file is part of MayaChemTools.

MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.

=cut
